File: 16-Proxy/src/go/real_image.go
package main

import "fmt"

// RealImage is the RealSubject — the expensive object. The costly load
// is simulated in the constructor.
type RealImage struct {
	filename string
}

func NewRealImage(filename string) *RealImage {
	img := &RealImage{filename: filename}
	img.loadFromDisk()
	return img
}

func (r *RealImage) loadFromDisk() {
	fmt.Printf("  [RealImage] Loading %s from disk (expensive)...\n", r.filename)
}

func (r *RealImage) Display() {
	fmt.Printf("  [RealImage] Displaying %s\n", r.filename)
}

File: 16-Proxy/src/rust/real_image.rs
use crate::image::Image;

/// RealSubject — the expensive object. The costly load is simulated in `new`.
pub struct RealImage {
    filename: String,
}

impl RealImage {
    pub fn new(filename: &str) -> Self {
        let image = RealImage { filename: filename.to_string() };
        image.load_from_disk();
        image
    }

    fn load_from_disk(&self) {
        println!("  [RealImage] Loading {} from disk (expensive)...", self.filename);
    }
}

impl Image for RealImage {
    fn display(&self) {
        println!("  [RealImage] Displaying {}", self.filename);
    }
}
